x-pack/solutions/observability/plugins/streams_app/public/hooks/queries/use_async_sample.tsx

```diff
@@ -0,0 +1,253 @@
+      documentSubscription.unsubscribe();
+      documentCountsSubscription.unsubscribe();
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
+    data.search,
+    convertedCondition,
+    options.start,
+    options.end,
+    options.size,
+    options.streamDefinition,
+    refreshId,
+  ]);
+
+  return {
+    isLoadingDocuments,
+    documentsError,
+    documents,
+    isLoadingDocumentCounts,
+    documentCountsError,
+    approximateMatchingPercentage,
+    refresh: () => setRefreshId((id) => id + 1),
+  };
+};
+
+export type AsyncSample = ReturnType<typeof useAsyncSample>;
+
+// Create runtime mappings for fields that aren't mapped.
+// Conditions could be using fields which are not indexed or they could use it with other types than they are eventually mapped as.
+// Because of this we can't rely on mapped fields to draw a sample, instead we need to use runtime fields to simulate what happens during
+// ingest in the painless condition checks.
+const getRuntimeMappings = (streamDefinition: WiredStreamGetResponse, condition: Condition) => {
+  const wiredMappedFields =
+    'wired' in streamDefinition.stream.ingest ? streamDefinition.stream.ingest.wired.fields : {};
+  const mappedFields = Object.keys(wiredMappedFields).concat(
+    Object.keys(streamDefinition.inherited_fields)
+  );
+
+  return Object.fromEntries(
+    getFields(condition)
+      .filter((field) => !mappedFields.includes(field.name))
+      .map((field) => [
+        field.name,
+        { type: field.type === 'string' ? 'keyword' : 'double' } as MappingRuntimeField,
+      ])
+  );
+};
+
+const getDocumentsSearchBody = (
+  condition: Condition,
+  options: Options,
+  runtimeMappings: MappingRuntimeFields
+) => {
+  const { size, start, end } = options;
+
+  const searchBody = {
+    query: {
+      bool: {
+        must: [
+          condition ? conditionToQueryDsl(condition) : { match_all: {} },
+          {
+            range: {
+              '@timestamp': {
+                gte: start,
+                lte: end,
+                format: 'epoch_millis',
+              },
+            },
+          },
+        ],
+      },
+    },
+    runtime_mappings: runtimeMappings,
+    sort: [
+      {
+        '@timestamp': {
+          order: 'desc',
+        },
+      },
+    ],
+    terminate_after: size,
+    track_total_hits: false,
+    size,
+  };
+  return searchBody;
+};
+
+const getDocumentCountsSearchBody = (
+  condition: Condition,
+  options: Options,
+  runtimeMappings: MappingRuntimeFields
+) => {
+  const { start, end } = options;
+
+  const searchBody = {
+    query: {
+      bool: {
+        must: [
+          {
+            range: {
+              '@timestamp': {
+                gte: start,
+                lte: end,
+                format: 'epoch_millis',
+              },
+            },
+          },
+        ],
+      },
+    },
+    aggs: {
+      sample: {
+        random_sampler: {
+          probability: 0.1,
```

Comment:
ideally we would get rid of this static sampling approach and build a two step dynamic sampler instead. 
My suggestion would be that we: 

count docs using just time filter 
if count > 100000 

probability = 100000 / docCount 

else
probability = 1

A result set of 100.000 docs should be enough for the user to get that validation (we can make it clear in the UI) 
Otherwise this 0.1 makes it only 10x faster. But this 10x will still be too slow if I have too many docs, so we need something more dynamic
Longer term the hope is that we'll get a function like this in ESQL, but we're not there yet

Reply:
++ (also you can't use a probability between 0.5 and 1, so we have to round up from 0.5)

Reply:
Thanks both. This was actually something I wanted to mention in the meeting today as I wasn't 100% sure on the best approach for the value, think you've covered it here 👌